### PR TITLE
Update train.jl to add a more detailed `train!` docstring

### PR DESCRIPTION
**Pull request by logankilpatrick**
Right now, the `train!` function docstring does not tell you what you need to do with respect to params. I tried to add that and moved a sentence from somewhere else in the docs about what the train function does to the top of the doc string.

### CODE REVIEW

File: src/optimise/train.jl [hunk 1]
```diff
@@ -82,11 +82,17 @@ batchmemaybe(x::Tuple) = x
...
+parameters (`params`) of your model based on a pluggable optimiser (`opt`).
+        
 For each datapoint `d` in `data`, compute the gradient of  `loss` with
 respect to `params` through backpropagation and call the optimizer `opt`.
-
+        
 If `d` is a tuple of arguments to `loss` call `loss(d...)`, else call `loss(d)`.
+        
+To get the trainable parameters, generally you pass the model to the [`Flux.params`](@ref)
+function (the just the layers you want to train) like `train!(loss, params(model), ...)`.
```
Comment (ToucheSir) [inline src/optimise/train.jl:95]:
```suggestion
To pass trainable parameters call [`Flux.params`](@ref) with your model or just the layers you want to train, like `train!(loss, params(model), ...)`.
```
Once https://github.com/FluxML/Flux.jl/pull/1786 lands, I think it makes sense to lean on that doc here.

File: src/optimise/train.jl [hunk 2]
```diff
@@ -82,11 +82,16 @@ batchmemaybe(x::Tuple) = x
...
     train!(loss, params, data, opt; cb)
-
+        
+`train!` uses a loss function (`loss`) and training data (`data`) to improve the 
+parameters (`params`) of your model based on a pluggable optimiser (`opt`).
+        
 For each datapoint `d` in `data`, compute the gradient of  `loss` with
 respect to `params` through backpropagation and call the optimizer `opt`.
-
+
```
Comment (ToucheSir) [inline src/optimise/train.jl:91]:
```suggestion

```

File: src/optimise/train.jl [hunk 3]
```diff
@@ -82,11 +82,16 @@ batchmemaybe(x::Tuple) = x
 
 """
     train!(loss, params, data, opt; cb)
-
+        
+`train!` uses a loss function (`loss`) and training data (`data`) to improve the 
+parameters (`params`) of your model based on a pluggable optimiser (`opt`).
+        
 For each datapoint `d` in `data`, compute the gradient of  `loss` with
```
Comment (ToucheSir) [inline src/optimise/train.jl:89]:
```suggestion
For each datapoint `d` in the `data` iterator, compute the gradient of  `loss` with
```
I don't love this phrasing, but barring a larger docstring refactor it seems the easiest way to point out what interfaces `data` should support.

File: src/optimise/train.jl [hunk 4]
```diff
@@ -82,11 +82,16 @@ batchmemaybe(x::Tuple) = x
 
 """
     train!(loss, params, data, opt; cb)
-
+        
+`train!` uses a loss function (`loss`) and training data (`data`) to improve the 
+parameters (`params`) of your model based on a pluggable optimiser (`opt`).
```
Comment (ToucheSir) [inline src/optimise/train.jl:87]: Is there any way for "optimiser" to link to https://github.com/FluxML/Flux.jl/blob/master/docs/src/training/optimisers.md?

File: src/optimise/train.jl (hunk 3 above)
Comment (logankilpatrick) [reply src/optimise/train.jl:89]: Do we need to say "iterator", isn't it assumed that any time we do `for something in something else`, `something else` is an iterator / iterable?

File: src/optimise/train.jl (hunk 4 above)
Comment (logankilpatrick) [reply src/optimise/train.jl:87]: yes, then we should also link the loss function page as well

Comment (logankilpatrick) [reply src/optimise/train.jl:87]: Should be simple per: https://juliadocs.github.io/Documenter.jl/stable/man/guide/#Cross-Referencing, I can add both if desired

File: src/optimise/train.jl (hunk 3 above)
Comment (ToucheSir) [reply src/optimise/train.jl:89]: I'm not sure, but my inclination would be that "in" is sufficiently general verbiage that new users wouldn't pattern match it to iteration. We can always leave this out and revisit if people aren't getting it.

**Review by ToucheSir — APPROVED** [review]

File: src/optimise/train.jl (hunk 4 above)
Comment (ToucheSir) [reply src/optimise/train.jl:87]: That would be great :)

**Comment by logankilpatrick** [comment]
Preview the PR here: https://fluxml.ai/Flux.jl/previews/PR1785/

**Comment by ToucheSir** [comment]
First 2 links to loss functions and datasets aren't resolving :confused:

**Comment by logankilpatrick** [comment]
Yeah I don't get why : ( I will open a issue for documenter.jl, I am sure I messed it up

**Comment by logankilpatrick** [comment]
Same here @ToucheSir I am going to try and tackle all of these doc ref issues in a larger PR since I will need to add ID's and such

File: src/optimise/train.jl [hunk 5]
```diff
@@ -82,16 +82,21 @@ batchmemaybe(x::Tuple) = x
 
 """
     train!(loss, params, data, opt; cb)
-
+        
+`train!` uses a [Loss Functions](@ref) (`loss`) and training [Datasets](@ref) (`data`) to improve the
```
Comment (ToucheSir) [inline src/optimise/train.jl:86]:
```suggestion
`train!` uses a `loss` function and training `data` to improve the 
```
Let's remove the dangling links for now so readers aren't confused by them.

**Review by ToucheSir — APPROVED** [review]
